docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,197 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As more custom attestation plugins are developed and more features are requested, new APIs are added to:
+
+1. Query registered attesters and verifiers.
+2. Select a format of the evidence that the registered attesters can generate.
+
+The following API was affected:
+
+1. Built-in SGX attester and verifier plugin
+
+This document describes the added and affected APIs as well as keeping
+the existing support for version 2 of the API.  This document assumes the user is familiar with
+the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. In order to simplify evidence generation in attestation applications, change was made to split built-in SGX attester into 2 attesters - one for SGX report and the other for SGX quote. Similar plugin splitting applies to built-in SGX verifier.
+
+Local and remote attestation is a TEE-specific feature. Specifying LOCAL or REMOTE flag when generating evidence can be confusing to many attesters. These flags were deprecated.
+
+User Experience
+---------------
+
+The user experience is improved by the ability of querying what the registered attestation verifiers can verify and attesters can attest. This is a typical attestation scenario:
+
+- Verifier application
```

Comment:
Do you foresee the opposite scenario happening as well? (i.e. attestor gives list of format id to verifier, which then picks one to verify or returns an error none match. Verifier returns the id to the atteestor, who then generates the evidence.)

Reply:
Yes. That is one possible scenario.

Another scenario is for attester to start sending evidence one by one until verifier accepts. The SDK supports many possible protocols.

---
In reply to: [386734837](https://github.com/openenclave/openenclave/pull/2621#discussion_r386734837) [](ancestors = 386734837)

Reply:
I think the flow between an attester and a verifier can be seen as bi-directional. Either side can propose or counter-propose. But eventually, the verifier is the entity that makes the final call whether it trusts the attester or not.

Reply:
Agree.

---
In reply to: [387202211](https://github.com/openenclave/openenclave/pull/2621#discussion_r387202211) [](ancestors = 387202211)

Reply:
Scenario 2 added.

---
In reply to: [386740316](https://github.com/openenclave/openenclave/pull/2621#discussion_r386740316) [](ancestors = 386740316,386734837)

Reply:
User experience scenario 2 added.

---
In reply to: [387207872](https://github.com/openenclave/openenclave/pull/2621#discussion_r387207872) [](ancestors = 387207872,387202211)